Add warnings for wrong values of model_confidence parameter (#8014)
RasaHQ/rasa: docs/docs/migration-guide.mdx

This page contains information about changes between major versions and
how you can migrate from one version to another.

## Rasa 2.3.3 to Rasa 2.3.4

:::caution

> This is a release **breaking backwards compatibility of machine learning models**.

Reviewer: @dakshvar22 Couldn't we have made that one non-breaking? Seem less than ideal to do a model breaking change  in a micro release.

Author: @wochinge It wasn't really possible to do in a non-breaking way because that would have meant leaving the bug around for all `2.3.x` versions. Here's the [slack](https://rasa-hq.slack.com/archives/C36SS4N8M/p1614079756041500) thread about it.

This page contains information about changes between major versions and
how you can migrate from one version to another.

## Rasa 2.3.3 to Rasa 2.3.4

:::caution
This is a release **breaking backwards compatibility of machine learning models**.
It is not possible to load previously trained models if they were trained with `model_confidence=cosine` or
`model_confidence=inner` setting. Please make sure to retrain the assistant before trying to use it with this improved version.

:::

### Machine Learning Components

> Rasa Open Source `2.3.0` introduced the option of using cosine similarities for model confidences by setting
> `model_confidence=cosine`. Some post-release experiments revealed that doing so can change the order of predicted
> labels. Since this is wrong, the option `model_confidence=cosine` has been removed and is no longer available.

Author: ```suggestion
Rasa Open Source `2.3.0` introduced the option of using cosine similarities for model confidences by setting
`model_confidence=cosine`. Some post-release experiments revealed that using `model_confidence=cosine` is wrong as it can change the order of predicted labels. That's why this option was removed in Rasa Open Source version `2.3.4`.
```

> Also, `model_confidence=inner` is deprecated as it produces an unbounded range of confidences which can break

Reviewer 2: ```suggestion
`model_confidence=inner` is deprecated as it produces an unbounded range of confidences which can break
```

Reviewer 2: why is this marked as resolved?

Reviewer 2: why is this marked as resolved?

Author: I think it got resolved because it got outdated. But I removed `Also` now.

the logic of assistants in various other places. Users are now encouraged to try `model_confidence=linear_norm`
which will produce a linearly normalized version of dot product similarities with
each value in the range `[0,1]`. Setting `model_confidence` to `linear_norm` will be helpful when trying to tune
fallback thresholds.

If the assistant was trained with `model_confidence=cosine` or `model_confidence=inner` setting, it should be
re-trained by either removing the `model_confidence` option from the configuration or setting it to `linear_norm`, as:
```
- name: DIETClassifier
  model_confidence: linear_norm
  constrain_similarities: True

>   ...

Reviewer 2: same edits to this whole thing as the comments I made on the changelog

> the logic of assistants in various other places. Users are now encouraged to try `model_confidence=linear_norm`
> which will produce a linearly normalized version of dot product similarities with
> each value in the range `[0,1]`. Setting `model_confidence` to `linear_norm` will be helpful when trying to tune
> fallback thresholds.
> 
> If the assistant was trained with `model_confidence=cosine` or `model_confidence=inner` setting, it should be
> re-trained by either removing the `model_confidence` option from the configuration or setting it to `linear_norm`, as:
> ```
> - name: DIETClassifier
>   model_confidence: linear_norm
>   constrain_similarities: True
>   ...

Author: ```suggestion
the logic of assistants in various other places. 

We encourage you to try `model_confidence=linear_norm` which will produce a linearly normalized version of dot product similarities with each value in the range `[0,1]`. This can be done with the following config:
```
- name: DIETClassifier
   model_confidence: linear_norm
   constrain_similarities: True
```

If you trained a model with `model_confidence=cosine` or `model_confidence=inner` setting using previous versions of Rasa Open Source, please re-train by either removing the `model_confidence` option from the configuration or setting it to `linear_norm`.
```

2. The default loss function (`loss_type=cross_entropy`) can add an optional sigmoid cross-entropy loss of all similarity values to constrain
them to an approximate range. You can turn on this option by setting `constrain_similarities=True`. This should help the models to perform better on real world test sets.

Also, a new option `model_confidence` has been added to `DIETClassifier` and `ResponseSelector`. It affects how the model's confidence for each label is computed during inference. It can take one of two values:
1. `softmax` - Similarities between input and label embeddings are post-processed with a softmax function, as a result of which confidence for all labels sum up to 1.

> 2. `inner` - Dot product similarity between input and label embeddings. Confidence for each label will be in an unbounded range.

Author: I've removed `cosine` as an option here so that any user coming to the migration guide from version `2.2`, does not see it as an option altogether. Is that okay?

Reviewer 2: yeah i think that makes sense

them to an approximate range. You can turn on this option by setting `constrain_similarities=True`. This should help the models to perform better on real world test sets.

Also, a new option `model_confidence` has been added to each ML component. It affects how the model's confidence for each label is computed during inference. It can take one of two values:
1. `softmax` - Dot product similarities between input and label embeddings are post-processed with a softmax function, as a result of which confidence for all labels sum up to 1.
2. `linear_norm` - Dot product similarities between input and label embeddings are post-processed with a linear normalization function. Confidence for each label will be in the range `[0,1]`.
The default value is `softmax`, but we recommend trying `linear_norm` as it should be easier to tune thresholds for fallback with it.
The value of this option does not affect how confidences are computed for entity predictions in `DIETClassifier`.

With both the above recommendations, users are encouraged to experiment by configuring their ML component, e.g. `DIETClassifier`, as:
```
- name: DIETClassifier

>   model_confidence: linear_norm

Reviewer 2: similar edits in this paragraph - but I don't think we should take out all reference to cosine here. We should just add a warning that the below is no longer true?

Author: Alright, I added back the possibility of adding `cosine` but changed the recommendation to `linear_norm`. Added an `EDIT` at the end about the deprecation.
